File: advent-of-code/haskell/app/Y2023D01.hs
module Y2023D01 ( year2023day01 ) where

import qualified Utils

import Data.Char (digitToInt, isDigit)
import Data.List (isPrefixOf)

filterString :: String -> String
filterString = filter isDigit

replaceAlphaNumbers :: String -> String
replaceAlphaNumbers [] = []
replaceAlphaNumbers str@(x:xs)
    -- drop one letter fewer than the length of the matching string to account
    -- for overlaps like "oneeight", "threeight" etc.
    | "one"   `isPrefixOf` str = '1' : replaceAlphaNumbers (drop 2 str)
    | "two"   `isPrefixOf` str = '2' : replaceAlphaNumbers (drop 2 str)
    | "three" `isPrefixOf` str = '3' : replaceAlphaNumbers (drop 4 str)
    | "four"  `isPrefixOf` str = '4' : replaceAlphaNumbers (drop 3 str)
    | "five"  `isPrefixOf` str = '5' : replaceAlphaNumbers (drop 3 str)
    | "six"   `isPrefixOf` str = '6' : replaceAlphaNumbers (drop 2 str)
    | "seven" `isPrefixOf` str = '7' : replaceAlphaNumbers (drop 4 str)
    | "eight" `isPrefixOf` str = '8' : replaceAlphaNumbers (drop 4 str)
    | "nine"  `isPrefixOf` str = '9' : replaceAlphaNumbers (drop 3 str)
    | otherwise                =  x  : replaceAlphaNumbers xs

decodeLine :: String -> Int
decodeLine str = 10 * digitToInt (head fstr) + digitToInt (last fstr)
    where
        fstr = filterString str

year2023day01 :: IO ()
year2023day01 = do
    list <- lines <$> readFile (Utils.getInputFile 2023 1)
    let
        part1 = sum [decodeLine str | str <- list]
        part2 = sum [decodeLine $ replaceAlphaNumbers str | str <- list]

    putStrLn $ Utils.resultString 2023 1 part1 part2

File: advent-of-code/jl/src/Y2023D01.jl
# src/Y2023D01.jl
# Simon Hulse
# simonhulse@protonmail.com
# Last Edited: Mon 09 Sep 2024 11:44:22 PM EDT

module Y2023D01

using Printf

export p₁, p₂

function process_number_words(line::AbstractString)::AbstractString
    # First deal with overlapping cases
    line = replace(line,
        "twone" => "21",
        "oneight" => "18",
        "threeight" => "38",
        "fiveight" => "58",
        "sevenine" => "79",
        "eightwo" => "82",
        "eighthree" => "83")
    # Now deal with isolated numbers
    line = replace(line,
        "one" => "1",
        "two" => "2",
        "three" => "3",
        "four" => "4",
        "five" => "5",
        "six" => "6",
        "seven" => "7",
        "eight" => "8",
        "nine" => "9")
    line
end

firstlast(s::AbstractString) = s[1] * s[lastindex(s)]

function run(input::AbstractString, process_num_words::Bool)::Integer
    func = process_num_words ? process_number_words : identity
    [
        parse(Int, firstlast(filter(isdigit, String(func(line)))))
        for line in split(input, "\n", keepempty=false)
    ] |> sum
end

p₁(input) = run(input, false)
p₂(input) = run(input, true)

end
